### PR TITLE
add logic to ensure control plane operator and hostedconfigoperator a…

### PR DESCRIPTION
**Pull request by relyt0925**
…re at expected version before reconciling a release update

This solves race conditions when upgrading a cluster (mostly across minor versions) that can result in the operator having orphaned resources. The operators will now nop when they aren't at the expected image version for a given release payload. Ultimately, they will be recreated, detect that they match the expected version, and then begin the reconciliation process. Solves: https://github.com/openshift/hypershift/issues/945

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes # https://github.com/openshift/hypershift/issues/945

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [x] This change includes unit tests.

### CODE REVIEW

File: support/util/imagefetch.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+	"context"
+	"fmt"
+	"github.com/blang/semver/v4"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/releaseinfo"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// GetHypershiftComponentImage resolves the appropriate control plane operator
```
Comment (enxebre) [inline support/util/imagefetch.go:13]: >// GetHypershiftComponentImage resolves the appropriate control plane operator
// image

So Rename GetHypershiftComponentImage -> GetCPOImage?

File: support/util/imagefetch.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	if hypershiftImage, exists := releaseInfo.ComponentImages()["hypershift"]; exists {
+		return hypershiftImage, nil
+	}
+
+	versionMajMin := fmt.Sprintf("%d.%d", version.Major, version.Minor)
+	switch versionMajMin {
+	case "4.9", "4.10":
+		return hypershiftOperatorImage, nil
```
Comment (enxebre) [inline support/util/imagefetch.go:44]: 4.10 should be always coming from the payload, i.e `cpoImage(HC.spec.release)`

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 3]
```diff
@@ -165,6 +166,24 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
...
+	pullSecret := common.PullSecret(hostedControlPlane.Namespace)
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(pullSecret), pullSecret); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to lookup pull secret: %w", err)
+	}
+	expectedImage, err := cpoutil.GetHypershiftComponentImage(ctx, hostedControlPlane.Annotations, hostedControlPlane.Spec.ReleaseImage, r.ReleaseProvider, hostedControlPlane.Annotations[hyperv1.ActiveHypershiftOperatorImage], pullSecret.Data[corev1.DockerConfigJsonKey])
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to lookup expected image: %w", err)
+	}
+	if expectedImage != r.ActiveImage {
+		r.Log.Info("No-oping until component is recreated with proper image", "expectedImage", expectedImage, "activeImage", r.ActiveImage)
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:183]: Is it safe to abruptly short-circuit like this? i.e wouldn't we need to check the HCP.Status as well to validate it's stable and no upgrades in-place so it's safe to short-circuit?

File: control-plane-operator/hostedclusterconfigoperator/controllers/common.go [hunk 4]
```diff
@@ -34,3 +41,21 @@ func nameMapper(names []string) handler.MapFunc {
 func NamedResourceHandler(names ...string) handler.EventHandler {
 	return handler.EnqueueRequestsFromMapFunc(nameMapper(names))
 }
+
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/common.go:44]: why do we have this absatraction here over GetHypershiftComponentImagebu t we don't in the HostedControlPlaneReconciler?

Also naming seems confusing, both funcs are achieving the same single goal, return a CPO image:
`getCPOImage`

This abstraction is enforcing a type to infer from in the signature, so if we want to keep it, it could be named:
`getExpectedCPOImageFromHCP`

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 5]
```diff
@@ -143,6 +146,14 @@ func (r *reconciler) reconcile(ctx context.Context) error {
 		return fmt.Errorf("failed to get hosted control plane %s/%s: %w", r.hcpNamespace, r.hcpName, err)
 	}
 
+	expectedImage, err := controllers.GetExpectedOperatorImage(ctx, r.cpClient, r.releaseProvider, hcp)
+	if err != nil {
+		return fmt.Errorf("failed to get expected image: %w", err)
+	}
+	if expectedImage != r.activeImage {
+		log.Info("No-oping until recreated with expected image", "activeImage", r.activeImage, "expectedImage", expectedImage)
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:154]: Is it safe to abruptly short-circuit like this? i.e wouldn't we need to check the HCP.Status as well to validate it's stable and no upgrades in-place so it's safe to short-circuit?

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 6]
```diff
@@ -165,6 +166,24 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 		}
 	}
 
+	// Determine if active control plane operator image matches the expected release image payload. If it doesn't
+	// need to no-op until the operator is recreated and at the matching version
+	if _, ok := hostedControlPlane.Annotations[hyperv1.ActiveHypershiftOperatorImage]; !ok {
```
Comment (alvaroaleman) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:171]: As mentioned in slack, this approach of checking inside the controllers is subject to time-of-check to time-of-us races when we are in an active reconciliation while this changes. Can you please explain why you went with this approach overmoving the CPO deployment rollout before the HCP reconciliation and wait for the CPO rollout to finish?

File: support/util/imagefetch_test.go [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+import (
+	"context"
+	"github.com/onsi/gomega"
+	hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+	"github.com/openshift/hypershift/support/releaseinfo"
+	fakereleaseprovider "github.com/openshift/hypershift/support/releaseinfo/fake"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/fake"
```
Comment (alvaroaleman) [inline support/util/imagefetch_test.go:12]: Can we please stick to the controller-runtime client to not pick up 20k LOC of dependencies and be consistent?

Comment (relyt0925) [reply support/util/imagefetch_test.go:12]: It's just testing dependencies on the fake client and some of the systems use client-go directly so I am hesitant to do that

Comment (relyt0925) [reply support/util/imagefetch_test.go:12]: I think it's beneficial to have that anyways as long as we are using client-go in the project which is used at a large scale.

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 6 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:171]: It's not subjected to time of check races with leader election:

This is mainly due to the fact now the update of the HostedCluster is atomic in the fact that it updates both the releaseImage and the expectedOperatorImage at a given point in time. 

That allows when a control-plane-operator picks up the update to see both changes at the same time and know if it matches the appropriate version. If not it nops below.

Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:171]: I don't want to introduce cross dependencies in the hypershift operator having to do advance introspection into the hostedcluster namespace (that isn't going to scale as this single entity manages up to 440 individual clusters in a single control plane). That's the main reason along with having a clean split of logic.

I guess the potential that you are pointing out is if somehow during the process leader election was lost, transferred to someone else, but yet this component kept running? I'm not sure if that's really something specific to this entity and most controllers would be subject to the same scenario in that case.

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 3 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:183]: It's safe and should occur:  We should not be updating resources until we are at the appropriate matching version. The one caveat is delete since at that point the hypershift operator will no longer reconcile the control plane operator. So the delete path should be allowed to continue as is.

File: control-plane-operator/hostedclusterconfigoperator/controllers/common.go (hunk 4 above)
Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/controllers/common.go:44]: HostedConfigOperator is setup to have multiple smaller controllers: So this would give an easy to use function that any controller needing to have this gating behavior to pull: Example of what I mean:
Controller 1: https://github.com/openshift/hypershift/blob/main/control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go
Controller 2: https://github.com/openshift/hypershift/blob/main/control-plane-operator/hostedclusterconfigoperator/controllers/cmca/setup.go

That being said: tangibly right now there is only the need to gate the one that is doing active reconciliation of cluster components. But wanted to provide something for the future to make it easy to do the same thing. I don't believe CPO image is the proper term as it's really HypershiftComponents (has cpo, hypershift, and hosted-cluster-config-operator) hence the name.


Am not opposed to renaming though but those are my thoughts

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go (hunk 5 above)
Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:154]: yep and is necessary to short circuit. Same explanation as above: but the only caveat is the delete path needs to still be allowed because at that point higher level components do not reconcile anymore. If it wasn't allowed it would be a deadlock.

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 6 above)
Comment (alvaroaleman) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:171]: It is, because an active reconciliation could be running while the version in the HCP gets changed. Leaderelection doesn't change anything about that, it just prevents two copies of the HCO running at the same time.

There is no advanced checking needed, we can find out through the deployment status if the latest version got rolled out (spec.Replics == status.readyReplicas == status.replicas == status.updatedReplicas)

File: support/util/imagefetch.go (hunk 1 above)
Comment (relyt0925) [reply support/util/imagefetch.go:13]: I should change the comment it's really not control-plane-operator the image holds multiple operator images in one

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 6 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:171]: at the start of the function here:
[https://github.com/openshift/hypershift/blob/7637a9237e2dccdfbcb35499ab02827ea317b[…]controllers/hostedcontrolplane/hostedcontrolplane_controller.go](https://github.com/openshift/hypershift/blob/7637a9237e2dccdfbcb35499ab02827ea317b178/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go#L138)

Let’s say after that the HC changes: that doesn’t matter because it’s not going to get picked up in that reconcilation loop


It’s already loaded it, then the next reconciliation loop when it gets the data again: It would see the updated data and see the check at the start and then start noping until a leader elected component had the proper image.

File: support/util/imagefetch.go (hunk 2 above)
Comment (relyt0925) [reply support/util/imagefetch.go:44]: I think people will still want this as a test path

Comment (relyt0925) [reply support/util/imagefetch.go:44]: so I do feel like the precedence is beneficial

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 6 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:171]: So let’s take your example a component in active reconciliation

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [27 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643992442662849?thread_ts=1643992413.467429&cid=C01C8502FMM)
It has already loaded it’s version of the hcp that it is “reconciling” locally

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [26 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643992500412159?thread_ts=1643992413.467429&cid=C01C8502FMM)
at the start of the function here:
[https://github.com/openshift/hypershift/blob/7637a9237e2dccdfbcb35499ab02827ea317b[…]controllers/hostedcontrolplane/hostedcontrolplane_controller.go](https://github.com/openshift/hypershift/blob/7637a9237e2dccdfbcb35499ab02827ea317b178/control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go#L138)
Let’s say after that the HC changes: that doesn’t matter because it’s not going to get picked up in that reconcilation loop

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [25 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643992554015949?thread_ts=1643992413.467429&cid=C01C8502FMM)
It’s already loaded it, then the next reconciliation loop when it gets the data again: It would see the updated data and see the check at the start and then start noping until a leader elected component had the proper image.

[Alvaro Aleman](https://app.slack.com/team/UT4JLAZHQ)  [19 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643992901879319?thread_ts=1643992413.467429&cid=C01C8502FMM)
Okay so everything the CPO could be doing wrong in a version mismatch case also requires the observed HCP to be in a version that differs?

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [16 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643993060607149?thread_ts=1643992413.467429&cid=C01C8502FMM)
Right yes it all depends on the data it’s seeing locally

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [16 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643993085497369?thread_ts=1643992413.467429&cid=C01C8502FMM)
This mainly was for my knowledge because I was concerned concurrent reconcile loops could be going on in the same component :slightly_smiling_face:

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [16 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643993103086459?thread_ts=1643992413.467429&cid=C01C8502FMM)
but I guess even then we would be covered because when it has the data

[Alvaro Aleman](https://app.slack.com/team/UT4JLAZHQ)  [9 minutes ago](https://coreos.slack.com/archives/C01C8502FMM/p1643993490363929?thread_ts=1643992413.467429&cid=C01C8502FMM)
No, the controller machinery only allows one concurrent reconcile per distinct key

[Tyler Lisowski](https://app.slack.com/team/U015GA40J4E)  [< 1 minute ago](https://coreos.slack.com/archives/C01C8502FMM/p1643994068925949?thread_ts=1643992413.467429&cid=C01C8502FMM)
will copy this slack convo to get too just thought it be easier for us to chat here real quick

File: support/util/imagefetch.go (hunk 1 above)
Comment (relyt0925) [reply support/util/imagefetch.go]: comment changed

File: support/util/imagefetch.go (hunk 2 above)
Comment (relyt0925) [reply support/util/imagefetch.go:44]: so for now my preference is to leave this as is for same behavior and a future pr can specifically remove the 4.10 piece if that is desired

**Comment by relyt0925** [comment]
Ready for second round of reviews

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 8]
```diff
@@ -165,6 +166,24 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
...
+		return ctrl.Result{}, fmt.Errorf("active hypershift operator image annotation not present")
+	}
+	pullSecret := common.PullSecret(hostedControlPlane.Namespace)
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(pullSecret), pullSecret); err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to lookup pull secret: %w", err)
+	}
+	expectedImage, err := cpoutil.GetHypershiftComponentImage(ctx, hostedControlPlane.Annotations, hostedControlPlane.Spec.ReleaseImage, r.ReleaseProvider, hostedControlPlane.Annotations[hyperv1.ActiveHypershiftOperatorImage], pullSecret.Data[corev1.DockerConfigJsonKey])
```
Comment (alvaroaleman) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:178]: Why do we need to do this rather than just checking if our own image matches what is in the annotation?

File: support/util/imagefetch.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	default:
+		return "", fmt.Errorf("unsupported release image with version %s", versionMajMin)
+	}
+}
+
+// LookupActiveContainerImage returns the image for a specified container in a specified pod
+func LookupActiveContainerImage(ctx context.Context, pods v1.PodInterface, podName string, containerName string) (string, error) {
```
Comment (alvaroaleman) [inline support/util/imagefetch.go:51]: Please use the controller runtime client like we do everywhere else in order to be consistent and avoid an additional ~20k LOC of deps. It will by default create a watch for all pods which we don't want, but we can set up a filter so that it will only watch its own pod, which is actually more efficient than making api requests here

Comment (ironcladlou) [reply support/util/imagefetch.go:51]: If https://github.com/openshift/hypershift/pull/963#discussion_r801840526 is possible, doesn't this all go away?

Comment (ironcladlou) [reply support/util/imagefetch.go:51]: I think maybe @alvaroaleman already mentioned this, could this be downward API?

Comment (ironcladlou) [reply support/util/imagefetch.go:51]: Downward API can't expose the container image directly, but it could inject from an annotation or label, which might be worth it considering the alternative (the client stuff here)?

Comment (relyt0925) [reply support/util/imagefetch.go:51]: No the dependency is added because we need the fake client set for testing of the standard client-go Kube Client. That is already utilized throughout the code and this provides a way to unit test it. If we want it gone: we would need to remove all references of that

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 8 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:178]: Not all HCPs specify the annotation: The algorithm is defined in the imageFetch support phase

Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:178]: I can refactor to where basically that annotation always exists on the HCP: and it's not a direct copy of the one on the HostedCluster though and instead that is all done at hostedCluster time when setting the label on the HCP. And then the HCP effectively always has the label.

File: support/util/imagefetch.go (hunk 9 above)
Comment (ironcladlou) [reply support/util/imagefetch.go:51]: FYI pod lookup will break the native process workflows (I don't use them anymore but others might)

File: control-plane-operator/main.go [hunk 10]
```diff
@@ -154,7 +154,11 @@ func NewStartCommand() *cobra.Command {
 			setupLog.Error(err, "unable to detect cluster capabilities")
 			os.Exit(1)
 		}
-
+		activeImage, err := util.LookupActiveContainerImage(context.TODO(), kubeClient.CoreV1().Pods(namespace), os.Getenv("POD_NAME"), "control-plane-operator")
```
Comment (ironcladlou) [inline control-plane-operator/main.go:157]: You have access to `ctx` here, should probably use it

Comment (ironcladlou) [reply control-plane-operator/main.go:157]: Might even want to make a subcontext with a timeout

File: control-plane-operator/hostedclusterconfigoperator/cmd.go [hunk 11]
```diff
@@ -208,6 +214,16 @@ func (o *HostedClusterConfigOperator) Run(ctx context.Context) error {
 			"konnectivity-agent": konnectivityAgentImage,
 		},
 	}
+	kubeClient, err := kubernetes.NewForConfig(cpConfig)
+	if err != nil {
+		setupLog.Error(err, "unable to create kube client")
+		os.Exit(1)
+	}
+	activeImage, err := util.LookupActiveContainerImage(context.TODO(), kubeClient.CoreV1().Pods(o.Namespace), os.Getenv("POD_NAME"), "hosted-cluster-config-operator")
```
Comment (ironcladlou) [inline control-plane-operator/hostedclusterconfigoperator/cmd.go:222]: Should probably use a timeout context based on `ctx`

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 12]
```diff
@@ -165,6 +166,17 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 		}
 	}
 
+	// Determine if the expected control plane operator image matches the active image in the control-plane-operator leader.
+	// If it doesn't need to no-op until the operator is recreated and at the matching version
+	if _, ok := hostedControlPlane.Annotations[hyperv1.ControlPlaneOperatorImageAnnotation]; !ok {
+		return ctrl.Result{}, fmt.Errorf("expected controlPlaneOperatorImageAnnotation not present on HCP")
+	}
+	expectedImage := hyperv1.ControlPlaneOperatorImageAnnotation
```
Comment (ironcladlou) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:174]: Not clear to me how this is working — `expectedImage` is being assigned to a constant... shouldn't this value come from the map lookup on L171? Seems like this is a mistake and is always `false`?

File: control-plane-operator/hostedclusterconfigoperator/cmd.go (hunk 11 above)
Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/cmd.go:222]: Agreed changing

Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/cmd.go:222]: for both control-plane-operator and hosted-cluster-config-operator

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 12 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:174]: it should be from the map you are correct

File: control-plane-operator/main.go [hunk 13]
```diff
@@ -154,7 +155,13 @@ func NewStartCommand() *cobra.Command {
 			setupLog.Error(err, "unable to detect cluster capabilities")
 			os.Exit(1)
 		}
-
+		lookupContext, lookupContextCancel := context.WithTimeout(ctx, 60*time.Second)
+		activeImage, err := util.LookupActiveContainerImage(lookupContext, kubeClient.CoreV1().Pods(namespace), os.Getenv("POD_NAME"), "control-plane-operator")
```
Comment (ironcladlou) [inline control-plane-operator/main.go:159]: @sjenning @cewong @enxebre @alvaroaleman this will break native process workflows, which I no longer use, but you might. An optional flag (so that we look up if possible and use the flag otherwise) could resolve it, but even in the latter case you would now have to pass a new flag w/ the image...

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go [hunk 14]
```diff
@@ -165,6 +166,17 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 		}
 	}
 
+	// Determine if the expected control plane operator image matches the active image in the control-plane-operator leader.
+	// If it doesn't need to no-op until the operator is recreated and at the matching version
+	if _, ok := hostedControlPlane.Annotations[hyperv1.ControlPlaneOperatorImageAnnotation]; !ok {
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:171]: Only need to access the index once:
```
expectedImage, ok := hostedControlPlane.Annotations[hyperv1.ControlPlaneOperatorImageAnnotation]
if !ok {
  return
}

if expectedImage != .ActiveImage {
```

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 15]
```diff
@@ -143,6 +144,16 @@ func (r *reconciler) reconcile(ctx context.Context) error {
 		return fmt.Errorf("failed to get hosted control plane %s/%s: %w", r.hcpNamespace, r.hcpName, err)
 	}
 
+	// Determine if the expected control plane operator image matches the active image in the hosted-cluster-config operator leader.
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:147]: Only need to access the index once:

expectedImage, ok := hostedControlPlane.Annotations[hyperv1.ControlPlaneOperatorImageAnnotation]
if !ok {
  return
}

if expectedImage != .ActiveImage {

Comment (enxebre) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:147]: If we end up moving forward with current approach, we'll want this to be a unit tested function e.g `isCPOCompatibleWithHCP()`

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 14 above)
Comment (enxebre) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go:171]: If we end up moving forward with current approach, we'll want this to be a unit tested function e.g `isCPOCompatibleWithHCP()`

File: support/util/imagefetch_test.go (hunk 7 above)
Comment (enxebre) [reply support/util/imagefetch_test.go:12]: I agree with @alvaroaleman, @relyt0925 I know we use the kube client directly in more places but we are heavily controller-runtime based and to continue towards consolidating the code base idiomatically across controllers in that direction.

Could we please just use mgr.GetAPIReader() to Lookup the active containerImage and also drop kube fake client dep?

Comment (relyt0925) [reply support/util/imagefetch_test.go:12]: I don't agree with that being in this initial refactor

Comment (relyt0925) [reply support/util/imagefetch_test.go:12]: I am fine with that being done in a future refactor that I can also drive but I don't want to keep introducing more changes on this one @enxebre .

Comment (relyt0925) [reply support/util/imagefetch_test.go:12]: actually scratch that that isn't bad I can do that

Comment (relyt0925) [reply support/util/imagefetch_test.go:12]: back to my original point: yea so I would like the refactor to be performed separate and it doesn't make sense in this initial iteration in my eyes.

I'm also happy to drive that one but I don't believe it should be done in this PR

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 8 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go]: done

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 16]
```diff
@@ -4,7 +4,7 @@ import (
 	"context"
 	"crypto/md5"
 	"fmt"
-
+	cpoutil "github.com/openshift/hypershift/support/util"
```
Comment (alvaroaleman) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:7]: Nit: Can we please keep the existing grouping intact and add this to the group below where all the other hypershift imports are?

File: control-plane-operator/hostedclusterconfigoperator/controllers/common.go [hunk 17]
```diff
@@ -1,11 +1,11 @@
 package controllers
 
 import (
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (alvaroaleman) [inline control-plane-operator/hostedclusterconfigoperator/controllers/common.go:4]: Nit: Please keep the grouping

File: control-plane-operator/hostedclusterconfigoperator/cmd.go [hunk 18]
```diff
@@ -16,8 +16,12 @@ import (
 	"errors"
 	"flag"
 	"fmt"
+	"github.com/openshift/hypershift/support/util"
```
Comment (alvaroaleman) [inline control-plane-operator/hostedclusterconfigoperator/cmd.go:19]: Nit: Importgrouping, please do not put third party imports into the stdlib group

File: support/util/cpocheck.go [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+import hyperv1 "github.com/openshift/hypershift/api/v1alpha1"
+
+// IsCPOCompatibleWithHCP checks to see if the active image in a control-plane-operator component matches the
+// expected image listed in the HCP object
+func IsCPOCompatibleWithHCP(hcpAnnotations map[string]string, activeImage string) bool {
+	expectedImage, ok := hcpAnnotations[hyperv1.ControlPlaneOperatorImageAnnotation]
+	if !ok {
+		return false
+	}
+	return expectedImage == activeImage
```
Comment (alvaroaleman) [inline support/util/cpocheck.go:12]: Nit: If it doesn't exist, `expectedImage` is an empty string, so you can just do
```suggestion
	return hcpAnnotations[hyperv1.ControlPlaneOperatorImageAnnotation] == activeIage
```

File: support/util/imagefetch.go [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+// preferred):
+//
+// 1. The image specified by the ControlPlaneOperatorImageAnnotation on the
+//    HostedControlPlane resource itself
+// 2. The hypershift image specified in the release payload indicated by the
+//    HostedControlPlane's release field
+// 3. The hypershift-operator's own image for release versions 4.9 and 4.10
+//
+// If no image can be found according to these rules, an error is returned.
+func GetHypershiftComponentImage(ctx context.Context, objectAnnotations map[string]string, releaseImage string, releaseProvider releaseinfo.Provider, hypershiftOperatorImage string, pullSecret []byte) (string, error) {
```
Comment (alvaroaleman) [inline support/util/imagefetch.go:24]: "ComponentImage" is the CPO image, right? Why was this moved around?

**Comment by enxebre** [comment]
The definition for `hypershift.openshift.io/control-plane-operator-image` is inconsistent now.

```
	// ControlPlaneOperatorImageAnnotation is a annotation that allows the specification of the control plane operator image.
	// This is used for development and e2e workflows
	ControlPlaneOperatorImageAnnotation = "hypershift.openshift.io/control-plane-operator-image"
```

 Could you please update it? Or actually I'm thinking we should probably have a different one for this since they have different semantics, e.g
```
ControlPlaneOperatorActiveImageAnnotation = "hypershift.openshift.io/control-plane-operator-active-image"
```

File: control-plane-operator/main.go (hunk 10 above)
Comment (relyt0925) [reply control-plane-operator/main.go]: done

File: control-plane-operator/hostedclusterconfigoperator/cmd.go (hunk 11 above)
Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/cmd.go]: done

File: control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go (hunk 14 above)
Comment (relyt0925) [reply control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go]: done

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go (hunk 15 above)
Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go]: done

File: support/util/imagefetch.go (hunk 20 above)
Comment (relyt0925) [reply support/util/imagefetch.go:24]: It was originally called by hosted-config-operator and the control-plane-operator so moved here. We can move it back since now it's just the hypershift operator that calls it

**Comment by relyt0925** [comment]
done @enxebre

File: support/util/cpocheck.go (hunk 19 above)
Comment (relyt0925) [reply support/util/cpocheck.go]: done

File: control-plane-operator/hostedclusterconfigoperator/cmd.go (hunk 18 above)
Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/cmd.go]: done

File: control-plane-operator/hostedclusterconfigoperator/controllers/common.go (hunk 17 above)
Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/controllers/common.go]: done

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go (hunk 16 above)
Comment (relyt0925) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go]: done

**Review by alvaroaleman — APPROVED** [review]
Thanks for going through all the trouble with this!

**Comment by relyt0925** [comment]
Thanks for all your time reviewing and advice @alvaroaleman and @enxebre and @sjenning and @ironcladlou

**Comment by relyt0925** [comment]
IBM Cloud E2E upgrade tests passed: merging
